packages/client/lib/miner/pendingBlock.ts

```diff
@@ -0,0 +1,171 @@
+  }
+
+  /**
+   * Starts building a pending block with the given payload
+   * @returns an 8-byte payload identifier to call {@link BlockBuilder.build} with
+   */
+  async start(vm: VM, parentBlock: Block, headerData: Partial<HeaderData> = {}) {
+    const number = parentBlock.header.number.addn(1)
+    const { gasLimit } = parentBlock.header
+    const baseFeePerGas = parentBlock.header.calcNextBaseFee()
+
+    // Set the state root to ensure the resulting state
+    // is based on the parent block's state
+    await vm.stateManager.setStateRoot(parentBlock.header.stateRoot)
+
+    const td = await vm.blockchain.getTotalDifficulty(parentBlock.hash())
+    vm._common.setHardforkByBlockNumber(parentBlock.header.number, td)
+
+    const builder = await vm.buildBlock({
+      parentBlock,
+      headerData: {
+        ...headerData,
+        number,
+        gasLimit,
+        baseFeePerGas,
+      },
+      blockOpts: {
+        putBlockIntoBlockchain: false,
+      },
+    })
+
+    const payloadId = randomBytes(8)
+    this.pendingPayloads.push([payloadId, builder])
+
+    // Add current txs in pool
+    const txs = await this.txPool.txsByPriceAndNonce(vm.stateManager, baseFeePerGas)
+    this.config.logger.info(
+      `Pending: Assembling block from ${txs.length} eligible txs (baseFee: ${baseFeePerGas})`
+    )
+    let index = 0
+    let blockFull = false
+    while (index < txs.length && !blockFull) {
+      try {
+        await builder.addTransaction(txs[index])
+      } catch (error: any) {
+        if (error.message === 'tx has a higher gas limit than the remaining gas in the block') {
+          if (builder.gasUsed.gt(gasLimit.subn(21000))) {
+            // If block has less than 21000 gas remaining, consider it full
+            blockFull = true
+            this.config.logger.info(
+              `Pending: Assembled block full (gasLeft: ${gasLimit.sub(builder.gasUsed)})`
+            )
+          }
+        } else {
+          // If there is an error adding a tx, it will be skipped
+          this.config.logger.debug(
+            `Pending: Skipping tx 0x${txs[index]
+              .hash()
+              .toString('hex')}, error encountered when trying to add tx:\n${error}`
+          )
+        }
+      }
+      index++
+    }
+
+    // Stop builder if block isn't returned after 1 minute
+    setTimeout(() => {
+      this.stop(payloadId)
+    }, 60000)
+
+    return payloadId
+  }
+
+  /**
+   * Stops a pending payload
+   */
+  stop(payloadId: Buffer) {
+    const payload = this.pendingPayloads.find((p) => p[0].equals(payloadId))
+    if (!payload) return
+    void payload[1].revert()
+  }
+
+  /**
+   * Returns the completed block
+   */
+  async build(payloadId: Buffer): Promise<Block | void> {
+    const payload = this.pendingPayloads.find((p) => p[0].equals(payloadId))
+    if (!payload) {
+      return
+    }
+    const builder = payload[1]
+
+    // Add new txs that the pool received
+    const txs = (
```

Comment:
Would it make sense to add a check like this from below
`if (builder.gasUsed.gt((builder as any).headerData.gasLimit.subn(21000)))` before you start looking in the transaction pool to add additional transactions?  That way, if the block is already full from the earlier call to `start`, you don't waste the time rummaging through the transaction pool.

Reply:
i think it's fine, this code is mostly unoptimized scaffolding to get some kind of basic functionality. this error would be thrown a few lines later that is handled appropriately.